### PR TITLE
Add benchmark

### PR DESCRIPTION
**Pull request by yasahi-hpc**
In this PR, I have added micro bench based on googlebenchmark. Need to implement more practical examples.

### CODE REVIEW

**Comment by yasahi-hpc** [comment]
Results on Icelake

```
Kokkos::OpenMP::initialize WARNING: OMP_PROC_BIND environment variable not set
  In general, for best performance with OpenMP 4.0 or better set OMP_PROC_BIND=spread and OMP_PLACES=threads
  For best performance with OpenMP 3.1 set OMP_PROC_BIND=true
  For unit testing set OMP_PROC_BIND=false

2024-01-31T21:58:11+09:00
Running ./fft/perf_test/_PerformanceTest_Benchmark
Run on (112 X 3400.04 MHz CPU s)
CPU Caches:
  L1 Data 32 KiB (x56)
  L1 Instruction 32 KiB (x56)
  L2 Unified 1024 KiB (x56)
  L3 Unified 39424 KiB (x2)
Load Average: 3.58, 3.96, 3.31
CPU architecture: none
Default Device: N6Kokkos6OpenMPE
GPU architecture: none
KOKKOSFFT_ENABLE_TPL_CUFFT: no
KOKKOSFFT_ENABLE_TPL_FFTW: yes
KOKKOSFFT_ENABLE_TPL_HIPFFT: no
KOKKOSFFT_ENABLE_TPL_ONEMKL: no
KOKKOS_COMPILER_INTEL_LLVM: 20230000
KOKKOS_ENABLE_ASM: yes
KOKKOS_ENABLE_CXX17: yes
KOKKOS_ENABLE_CXX20: no
KOKKOS_ENABLE_CXX23: no
KOKKOS_ENABLE_DEBUG_BOUNDS_CHECK: no
KOKKOS_ENABLE_HBWSPACE: no
KOKKOS_ENABLE_HWLOC: no
KOKKOS_ENABLE_INTEL_MM_ALLOC: no
KOKKOS_ENABLE_LIBDL: yes
KOKKOS_ENABLE_LIBRT: no
KOKKOS_ENABLE_OPENMP: yes
KOKKOS_ENABLE_PRAGMA_IVDEP: no
KOKKOS_ENABLE_PRAGMA_LOOPCOUNT: no
KOKKOS_ENABLE_PRAGMA_UNROLL: no
KOKKOS_ENABLE_PRAGMA_VECTOR: no
Kokkos: OpenMP thread_pool_topology[ 1 x 112 x 1 ]
Kokkos Version: 4.2.0
KokkosFFT Version: 0.0.0
platform: 64bit
------------------------------------------------------------------------------------------------------------------------
Benchmark                                                              Time             CPU   Iterations UserCounters...
------------------------------------------------------------------------------------------------------------------------
FFT_1DView<float, Kokkos::LayoutLeft>/N:4096/manual_time            5330 us         1775 us          129 GB/s=0.0122966/s MB (In)=0.032768 MB (Out)=0.032768
FFT_1DView<float, Kokkos::LayoutLeft>/N:8192/manual_time             882 us          876 us          597 GB/s=0.148661/s MB (In)=0.065536 MB (Out)=0.065536
FFT_1DView<float, Kokkos::LayoutLeft>/N:16384/manual_time            820 us          816 us          860 GB/s=0.31978/s MB (In)=0.131072 MB (Out)=0.131072
FFT_1DView<float, Kokkos::LayoutLeft>/N:32768/manual_time           2225 us         2209 us          343 GB/s=0.235583/s MB (In)=0.262144 MB (Out)=0.262144
FFT_1DView<float, Kokkos::LayoutLeft>/N:65536/manual_time           3881 us         3810 us          216 GB/s=0.270188/s MB (In)=0.524288 MB (Out)=0.524288
```

**Comment by yasahi-hpc** [comment]
Build errors on HIP

```
[ 78%] Building CXX object fft/perf_test/CMakeFiles/_PerformanceTest_Benchmark.dir/PerfTest_FFT1.cpp.o
clang-14: warning: argument unused during compilation: '-fno-gpu-rdc' [-Wunused-command-line-argument]
In file included from /work/fft/perf_test/PerfTest_FFT1.cpp:2:
In file included from /work/fft/perf_test/Benchmark_Context.hpp:9:
In file included from /work/tpls/kokkos/core/src/Kokkos_Core.hpp:43:
In file included from /work/tpls/kokkos/core/src/Kokkos_Core_fwd.hpp:28:
In file included from /work/tpls/kokkos/core/src/Kokkos_Macros.hpp:96:
In file included from /work/build_HIP/tpls/kokkos/KokkosCore_Config_SetupBackend.hpp:22:
In file included from /work/tpls/kokkos/core/src/setup/Kokkos_Setup_HIP.hpp:24:
In file included from /opt/rocm-5.2.0/hip/include/hip/hip_runtime.h:26:
/opt/rocm-5.2.0/hip/include/hip/../../../include/hip/hip_runtime.h:66:2: error: ("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
#error("Must define exactly one of __HIP_PLATFORM_AMD__ or __HIP_PLATFORM_NVIDIA__");
 ^
```

**Review by acalloo — APPROVED** [review]
Just two minor comments for my understanding.

File: CMakeLists.txt [hunk 1]
```diff
@@ -6,8 +6,21 @@ list(INSERT CMAKE_MODULE_PATH 0 "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
...
-option(KokkosFFT_ENABLE_HOST_AND_DEVICE "Enable FFT on both host and device")
-option(KokkosFFT_INTERNAL_Kokkos "Build internal Kokkos instead of relying on external one")
+option(KokkosFFT_ENABLE_HOST_AND_DEVICE "Enable FFT on both host and device" OFF)
+option(KokkosFFT_INTERNAL_Kokkos "Build internal Kokkos instead of relying on external one" OFF)
+option(KokkosFFT_ENABLE_BENCHMARK "Build benchmarks for KokkosFFT" OFF)
+
+# Version information
+set(KokkosFFT_VERSION_MAJOR 0)
+set(KokkosFFT_VERSION_MINOR 0)
+set(KokkosFFT_VERSION_PATCH 00)
```
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply CMakeLists.txt:16]: Not sure. As well as DDC, I am planning to begin with 0.0.00.

File: CMakeLists.txt [hunk 2]
```diff
@@ -6,8 +6,21 @@ list(INSERT CMAKE_MODULE_PATH 0 "${CMAKE_CURRENT_SOURCE_DIR}/cmake")
...
+set(KokkosFFT_VERSION "${KokkosFFT_VERSION_MAJOR}.${KokkosFFT_VERSION_MINOR}.${KokkosFFT_VERSION_PATCH}")
+
+#Set variables for config file
+math(EXPR KOKKOSFFT_VERSION "${KokkosFFT_VERSION_MAJOR} * 10000 + ${KokkosFFT_VERSION_MINOR} * 100 + ${KokkosFFT_VERSION_PATCH}")
+math(EXPR KOKKOSFFT_VERSION_MAJOR "${KOKKOSFFT_VERSION} / 10000")
+math(EXPR KOKKOSFFT_VERSION_MINOR "${KOKKOSFFT_VERSION} / 100 % 100")
+math(EXPR KOKKOSFFT_VERSION_PATCH "${KOKKOSFFT_VERSION} % 100")
```
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply CMakeLists.txt:23]: `KOKKOSFFT_VERSION` is a temporary variable to convert version variables into integers. What we would really like to have in integers are `KOKKOSFFT_VERSION_MAJOR`, `KOKKOSFFT_VERSION_MINOR` and `KOKKOSFFT_VERSION_PATCH`. These are used to replace the [placeholders](https://github.com/CExA-project/kokkos-fft/blob/df08e96c7867914c1fc1749819dc738146056dd7/cmake/KokkosFFT_config.h.in#L5-L8) in c++ codes that are used to show version info in benchmark.

 Is it OK for you?

**Comment by yasahi-hpc** [comment]
@acalloo First of all, thank you for your reviews.

**Comment by yasahi-hpc** [comment]
@pzehner I found the errors in hip build, which says

```
Traceback (most recent call last):
  File "/opt/rocm-5.2.0/bin/rocm_agent_enumerator", line 257, in <module>
    main()
  File "/opt/rocm-5.2.0/bin/rocm_agent_enumerator", line 241, in main
    target_list = readFromKFD()
  File "/opt/rocm-5.2.0/bin/rocm_agent_enumerator", line 193, in readFromKFD
    for node in sorted(os.listdir(topology_dir)):
FileNotFoundError: [Errno 2] No such file or directory: '/sys/class/kfd/kfd/topology/nodes/'
```

For some reason, build tests completed.

There seems to be a fix for that after rocm5.3.0+. 
I will try to update the base image for hip.
https://github.com/ROCm/rocminfo/issues/56

File: CMakeLists.txt (hunk 2 above)
[the remark this answers is not in the document]
Comment (cedricchevalier19) [reply CMakeLists.txt:23]: Why not use https://cmake.org/cmake/help/latest/variable/PROJECT_VERSION.html#variable:PROJECT_VERSION ?

[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply CMakeLists.txt:23]: As commented, this will be done in the fix of #32
For the moment, need to focus on docs and [thread capability](#31).

**Pull request merged**